apps/aecore/src/aec_peer_connection.erl

```diff
@@ -1277,22 +1277,48 @@ pre_assembly_check(MicroHeader) ->
     case aec_db:has_block(Hash) of
         true -> known;
         false ->
-            Validators = [fun validate_micro/1,
-                          fun validate_connected_to_chain/1,
-                          fun validate_delta_height/1,
-                          fun validate_prev_key_block/1,
-                          fun validate_micro_signature/1],
-            aeu_validation:run(Validators, [MicroHeader])
+            {PrevHeader, PrevKeyHeader} = get_onchain_env(MicroHeader),
+            Validators = [fun validate_micro_block_header/3,
+                          fun validate_connected_to_chain/3,
+                          fun validate_delta_height/3,
+                          fun validate_prev_key_block/3,
+                          fun validate_micro_signature/3],
+            aeu_validation:run(Validators, [MicroHeader, PrevHeader, PrevKeyHeader])
     end.
 
-validate_connected_to_chain(MicroHeader) ->
-    case aec_chain_state:hash_is_connected_to_genesis(
-            aec_headers:prev_hash(MicroHeader)) of
-        true -> ok;
+get_onchain_env(MicroHeader) ->
+    PrevHeaderHash = aec_headers:prev_hash(MicroHeader),
+    PrevKeyHash = aec_headers:prev_key_hash(MicroHeader),
+    case aec_chain:get_header(PrevHeaderHash) of
+        {ok, PrevHeader} ->
+            case aec_headers:type(PrevHeader) of
+                key ->
+                    {PrevHeader, PrevHeader};
+                micro ->
+                    case aec_chain:get_header(PrevKeyHash) of
+                        {ok, PrevKeyHeader} -> {PrevHeader, PrevKeyHeader};
+                        error               -> {PrevHeader, not_found}
```

Comment:
Yes, it could make sense. I'll merge this though and can create a new PR for that. It will also simplify the `validate_*` functions as I can remove the lines that are _virtually impossible_